rust/src/tests/coordinator.rs

```diff
@@ -74,3 +74,93 @@ async fn full_cycle_1_round_1_participant() {
         }
     }
 }
+
+#[tokio::test]
+async fn dropout_1_participant_during_training() {
+    let settings = FederatedLearningSettings {
+        rounds: 1,
+        participants_ratio: 1.0,
+        min_clients: 2,
+        heartbeat_timeout: 1,
+    };
+    let (rpc_client, service_handle, _join_handle) = start_service(settings);
+
+    // Create first client. Since min_clients is 2, the heartbeat response should be `StandBy`.
+    let id_1 = service_handle.rendez_vous_accepted().await;
+    let hb_resp = service_handle.heartbeat(id_1).await;
+    assert_eq!(hb_resp, HeartBeatResponse::StandBy);
+
+    // Create second client.
+    let id_2 = service_handle.rendez_vous_accepted().await;
+
+    // Now the the round can start because the requirement of min two clients is fulfilled.
+    let round = service_handle.heartbeat_selected(id_1).await;
+    assert_eq!(round, 0);
+    let round = service_handle.heartbeat_selected(id_2).await;
+    assert_eq!(round, 0);
+
+    // Both clients start the training.
+    rpc_client
+        .mock()
+        .expect_select()
+        .returning(|_, _| future::ready(Ok(())));
+    rpc_client
+        .mock()
+        .expect_select()
+        .returning(|_, _| future::ready(Ok(())));
+    let (url, _token) = service_handle.start_training_accepted(id_1).await;
+    assert_eq!(&url, AGGREGATOR_URL);
+    let (url, _token) = service_handle.start_training_accepted(id_2).await;
+    assert_eq!(&url, AGGREGATOR_URL);
+
+    // Here we reset only the heartbeat of the first client and wait until the
+    // heartbeat timeout of the second client has been reached.
+    service_handle.heartbeat(id_1).await;
+    sleep_ms(500).await;
+    service_handle.heartbeat(id_1).await;
+    sleep_ms(500).await;
+    service_handle.heartbeat(id_1).await;
+    // The second client should be dropped.
+
+    // The first client finished training.
+    service_handle.end_training(id_1, true).await;
+
+    // Create a third client. The third client should be selected by the coordinator and
+    // be able to participate in the training session.
+    let id_3 = service_handle.rendez_vous_accepted().await;
+    let round = service_handle.heartbeat_selected(id_3).await;
+    assert_eq!(round, 0);
+    rpc_client
+        .mock()
+        .expect_select()
+        .returning(|_, _| future::ready(Ok(())));
+    let (url, _token) = service_handle.start_training_accepted(id_3).await;
+    assert_eq!(&url, AGGREGATOR_URL);
+
+    // Let's simulate that the second client tries to reconnect to the coordinator.
+    let id_2 = service_handle.rendez_vous_accepted().await;
+
+    // The second client should be accepted but not re-selected (for current round)
+    // by the coordinator.
+    let hb_resp = service_handle.heartbeat(id_2).await;
+    assert_eq!(hb_resp, HeartBeatResponse::StandBy);
+
+    // The third client finished training.
+    service_handle.end_training(id_3, true).await;
+
+    // Trigger aggregation.
+    rpc_client
```

Comment:
I'm surprised this works: `service_handle.end_training` should trigger the aggregation, and since you only set `rpc_client.expect_aggregate()` _after_, I would expect this to fail. I'll do more tests on this after lunch if you don't mind, there might be something wrong with the mock.